src/DidCache.ts

```diff
@@ -1,16 +1,338 @@
+    }
+
+    return opHash;
+  }
+
+  public constructor (
+    private readonly cas: Cas,
+    private readonly didDocUpdate: DidDocumentUpdater,
+    private readonly didDocCreate: DidDocumentCreator) {
+
+  }
+
+  /**
+   * Rollback the state of the DidCache by removing all operations
+   * with transactionNumber greater than the provided parameter value.
+   * The intended use case for this method is to handle rollbacks
+   * in the blockchain.
+   *
+   * The current implementation is inefficient: It simply scans the two
+   * hashmaps storing the core Did state and removes all entries with
+   * a greater transaction number.  In future, the implementation should be optimized
+   * for the common case by keeping a sliding window of recent operations.
+   */
+  public rollback (transactionNumber: number) {
+
+    // Iterate over all nextVersion entries and remove all versions
+    // with "next" operation with transactionNumber greater than the provided
+    // parameter.
+    this.nextVersion.forEach((opHash, version, map) => {
+      const opInfo = this.opHashToInfo.get(opHash) as OperationInfo;
+      if (opInfo.transactionNumber > transactionNumber) {
+        map.delete(version);
+      }
+    });
+
+    // Iterate over all operations and remove those with with
+    // transactionNumber greater than the provided parameter.
+    this.opHashToInfo.forEach((opInfo, opHash, map) => {
+      if (opInfo.transactionNumber > transactionNumber) {
+        map.delete(opHash);
+      }
+    });
+  }
+
+  /**
+   * Resolve a did.
+   */
+  public async resolve (did: VersionId): Promise<DidDocument | null> {
+    const latestVersion = this.last(did);
+
+    // lastVersion === null implies we do not know about the did
+    if (latestVersion === null) {
+      return null;
+    }
+
+    return this.lookup(latestVersion);
+  }
+
+  /**
+   * Returns the Did document for a given version identifier.
+   */
+  public async lookup (versionId: VersionId): Promise<DidDocument | null> {
+    // Version id is also the operation hash that produces the document
+    const opHash = versionId;
+
+    const opInfo = this.opHashToInfo.get(opHash);
+
+    // We don't know anything about this operation
+    if (opInfo === undefined) {
+      return null;
+    }
+
+    // Construct the operation using a CAS lookup
+    const op = await this.getOperation(opInfo);
+
+    if (this.isInitialVersion(opInfo)) {
+      return this.didDocCreate(op);
+    } else {
+      const prevVersion = op.previousOperationHash as VersionId;
+      const prevDidDoc = await this.lookup(prevVersion);
+      if (prevDidDoc === null) {
+        return null;
+      } else {
+        return this.didDocUpdate(prevDidDoc, op);
+      }
+    }
+  }
+
+  /**
+   * Return the previous version id of a given DID version. The implementation
+   * is inefficient and involves an async cas read. This should not be a problem
+   * since this method is not hit for any of the externally exposed DID operations.
```

Comment:
Why not store the previous version also and be efficient since the reverse (ie next version) is already stored? If the reasoning is that the method is not being used currently, then I'd say just remove it.

Reply:
Related to the previous issue. If we do decide to keep it would be for purely testing purposes where performance is not a critical issue. Whether or not we retain prev/first methods, I think we should not keep a hashmap version->prevversion since we probably want the state of the lean cache (without yet to be added perf optimizations) as small as possible. Does this sound reasonable?